packages/headless/src/controllers/triggers/headless-redirection-trigger.ts

```diff
@@ -0,0 +1,106 @@
+import {Schema, Value} from '@coveo/bueno';
+import {Engine} from '../../app/headless-engine';
+import {
+  ConfigurationSection,
+  RedirectionSection,
+} from '../../state/state-sections';
+import {configuration, redirection} from '../../app/reducers';
+import {buildController, Controller} from '../controller/headless-controller';
+import {loadReducerError} from '../../utils/errors';
+import {validateOptions} from '../../utils/validate-payload';
+import {logTriggerRedirect} from '../../features/redirection/redirection-analytics-actions';
+
+export interface RedirectionTriggerOptions {
+  /**
+   * The function used to handle whenever the `RedirectionTrigger` controller's state's `redirectTo` value changes.
+   */
+  onRedirect(): void;
+}
+
+export interface RedirectionTriggerProps {
+  /**
+   * The options for the `RedirectionTrigger` controller.
+   */
+  options?: RedirectionTriggerOptions;
+}
+
+const optionsSchema = new Schema({
+  onRedirect: new Value({required: true}),
+});
+
+/**
+ * The `RedirectionTrigger` controller handles redirection actions.
+ */
+export interface RedirectionTrigger extends Controller {
+  /**
+   * the state of the `RedirectionTrigger` controller.
+   */
+  state: RedirectionTriggerState;
+}
+
+/**
+ * A scoped and simplified part of the headless state that is relevant to the `RedirectionTrigger` controller.
+ */
+export interface RedirectionTriggerState {
+  /**
+   * The url used for the redirection.
+   */
+  redirectTo: string;
+}
+
+/**
+ * Creates a `RedirectionTrigger` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `Redirection` properties.
+ * @returns A `Redirection` controller instance.
+ * */
+export function buildRedirection(
+  engine: Engine<object>,
+  props: RedirectionTriggerProps = {}
+): RedirectionTrigger {
+  if (!loadRedirectionReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+
+  const options = validateOptions(
+    engine,
+    optionsSchema,
+    props.options,
+    'buildRedirectionTrigger'
+  ) as Required<RedirectionTriggerOptions>;
+
+  const redirectTo = engine.state.redirection.redirectTo!;
+  const {onRedirect} = options;
+
+  return {
+    ...controller,
+
+    subscribe() {
+      const strictListener = () => {
+        if (redirectTo !== '') {
+          onRedirect();
+          dispatch(logTriggerRedirect);
+        }
+      };
+      strictListener();
+      return engine.subscribe(strictListener);
```

Comment:
This `subscribe` block does not look correct. Controllers usually do not need to implement `subscribe`. I'm also not clear on why calling it sends an analytics event, hmm

Reply:
We would like the handler function `onRedirect` passed down as an option to be called whenever the engine's state's `redirectTo` value changes. Whenever the handler is called, we also want to log the trigger to analytics. That is what I tried to achieve by overriding the `subscribe` method like in `headless-url-manager.ts` .

Reply:
Hmm, I think the flow of information is strictly in one direction, from the platform -> headless in this case. There's no need to pass information in? How would someone use the controller? Wouldn't be something like:

```
const trigger = buildTrigger(...)
trigger.subscribe(() => {
     window.location.href = trigger.state.redirectionUrl
})
```

Reply:
I'm not sure, I would have to see with Jean-Felix. I might have misunderstood but I thought I would extract the redirectionUrl from `engine.state.redirection.redirectTo`